bftengine/src/preprocessor/messages/PreProcessedResultMsg.cpp

```diff
@@ -0,0 +1,106 @@
+#include "Replica.hpp"  // for HAS_PRE_PROCESSED_FLAG
+#include "endianness.hpp"
+namespace preprocessor {
+
+PreProcessedResultMsg::PreProcessedResultMsg(NodeIdType sender,
+                                             uint64_t reqSeqNum,
+                                             uint32_t resultLength,
+                                             const char* result,
+                                             uint64_t reqTimeoutMilli,
+                                             const std::string& cid,
+                                             const concordUtils::SpanContext& spanContext,
+                                             const char* messageSignature,
+                                             uint32_t messageSignatureLen,
+                                             const char* resultSignaturesBuf,
+                                             const uint32_t resultSignaturesBufLen)
+    : ClientRequestMsg(sender,
+                       bftEngine::HAS_PRE_PROCESSED_FLAG,
+                       reqSeqNum,
+                       resultLength,  // check the comment in the header to
+                       result,        // understand why result is passed as request here
+                       reqTimeoutMilli,
+                       cid,
+                       spanContext,
+                       messageSignature,
+                       messageSignatureLen,
+                       resultSignaturesBufLen) {
+  msgBody_->msgType = MsgCode::PreProcessedResult;
+  // ClientRequestMsg allocates additional memory for the signatures
+  // Get pointer to it here and assert that the buffer is big enough
+  auto [pos, max_len] = getExtraBufPtr();
+  ConcordAssert(max_len >= resultSignaturesBufLen);
+
+  memcpy(pos, resultSignaturesBuf, resultSignaturesBufLen);
+}
+
+PreProcessedResultMsg::PreProcessedResultMsg(bftEngine::ClientRequestMsgHeader* body) : ClientRequestMsg(body) {}
+
+std::pair<char*, uint32_t> PreProcessedResultMsg::getResultSignaturesBuf() { return getExtraBufPtr(); }
+
+std::string PreProcessResultSignature::serializeResultSignatureList(
+    const std::list<PreProcessResultSignature>& signatures) {
+  size_t buf_len = 0;
+  for (const auto& s : signatures) {
+    buf_len += sizeof(s.sender_replica) + sizeof(uint32_t) + s.signature.size();
+  }
+
+  std::string output;
+  output.reserve(buf_len);
+
+  for (const auto& s : signatures) {
+    output.append(concordUtils::toBigEndianStringBuffer(s.sender_replica));
+    output.append(concordUtils::toBigEndianStringBuffer<uint32_t>(s.signature.size()));
+    output.append(s.signature.begin(), s.signature.end());
+  }
+
+  return output;
+}
+
+std::list<PreProcessResultSignature> PreProcessResultSignature::deserializeResultSignatureList(const char* buf,
+                                                                                               size_t len) {
+  size_t pos = 0;
+  std::list<PreProcessResultSignature> ret;
+  while (1) {
```

Comment:
We should know ahead how long is the signatures list. The signature size is also constant.

Reply:
I agree that we should know that there will be f signatures and its size is constant, but I have a reason to do it in such a way.

I put the signature lengths so that the message encoding/decoding is decoupled from the SIgManager. I think it's better to pass value sizes in the message unless they are a hard constant. In this case I don't consider them hard constant, because if we change the hashing algorithm this size will change. Replicas will easily figure this out, but the message will be unparsable outside Concord.

My motivation about the list size is similar - we can do for loop from 0 to F, but I wanted to implement more independant message format. I could have put the list size in the message too, but it seemed redundant for me.